navigation/visual_navigation_base.h

```diff
@@ -40,25 +40,25 @@ class VisualNavigationBase
     //------------------------------------------------------------------------
     // Public API
     //------------------------------------------------------------------------
-    //! Train with image at specified path
-    void train(const filesystem::path &imagePath, bool resizeImage = false)
-    {
-        if (!tryTrain(imagePath, resizeImage)) {
-            throw std::runtime_error("Path " + imagePath.str() + " does not exist");
-        }
-    }
-
     //! Train algorithm with specified route
-    void trainRoute(const filesystem::path &routePath, bool resizeImages = false)
+    void trainRoute(const ImageDatabase &imdb, bool resizeImages = false)
     {
-        if (!routePath.exists()) {
-            throw std::runtime_error("Path " + routePath.str() + " does not exist");
-        }
-
-        for (size_t i = 0;; i++) {
-            const auto filename = routePath / getRouteDatabaseFilename(i);
-            if (!tryTrain(filename, resizeImages)) {
-                return;
+        cv::Mat image;
+        if (resizeImages) {
+            cv::Mat imageResized;
+            for (const auto &e : imdb) {
+                image = e.loadGreyscale();
+                assert(image.type() == CV_8UC1);
```

Comment:
I think these asserts are unnecessary - the load greyscale function should return a greyscale image!

Reply:
True!

Reply:
...although on checking it, it's not the greyscale-ness we're checking here, but whether or not the image is of type uint8_t

Reply:
Fine point - ``cv::imread`` looks like it could theoretically return some HDR weirdness